### ISSUE TITLE
Support for Updated Air Purifier Models

### ISSUE THREAD

**Issue by lhttony** [Sunoo/homebridge-philips-air#1]
As I’ve just tried on my Philips AC4558 but no luck yet, thanks

**Comment by Sunoo** [comment]
Yes, it is. I developed it with an AC1214, but I don't see any reason yours wouldn't work. Are you getting any errors in your logs or anything? I can try to help get it working.

**Comment by lhttony** [comment]
Thanks Sunoo,
the log is as below, I've double checked that my Purifier IP is correct :

[2/14/2020, 11:59:51 AM] [philipsAir] Initializing platform accessory Air Purifier...
[2/14/2020, 11:59:52 AM] [philipsAir] Unable to load firmware info: FetchError: request to http://192.168.1.40/di/v1/products/0/firmware failed, reason: connect ECONNREFUSED 192.168.1.40:80
[2/14/2020, 11:59:53 AM] [philipsAir] Unable to load status info: FetchError: request to http://192.168.1.40/di/v1/products/1/air failed, reason: connect ECONNREFUSED 192.168.1.40:80
[2/14/2020, 11:59:54 AM] [philipsAir] Unable to load filter info: FetchError: request to http://192.168.1.40/di/v1/products/1/fltsts failed, reason: connect ECONNREFUSED 192.168.1.40:80
(node:18973) [DEP0025] DeprecationWarning: sys is deprecated. Use util instead.

**Comment by Sunoo** [comment]
Interesting. Maybe something _is_ different about the AC4558... I’ll do some looking online and see if I can find anything out.

In the meantime, can you try going to http://192.168.1.40/ in a web browser and let me know what you see?

**Comment by lhttony** [comment]
It's not responding to the http request :

This site can’t be reached
192.168.1.40 took too long to respond.

maybe such part is changed..

**Comment by Sunoo** [comment]
I did some looking, looks like the newer models have completely changed their APIs.

I could probably attempt to blindly code for it, but it’d be somewhat difficult not being able to test as I work.

**Comment by lhttony** [comment]
wow, that would be great and I can try it myself for the testing part,
but please take your time on this and thanks very much for your effort.

**Comment by Sunoo** [comment]
I’ll take a crack at it next time I’ve got some free time. No promises that it results in something working.

**Comment by lhttony** [comment]
No problem, thanks in advance

**Comment by Sunoo** [comment]
Just an update: Looks like people are still working on figuring out how to talk to the new models, but are making progress. Especially since I don't have one of these models myself, I'm going to be waiting on them to work it out.

**Comment by jamesg311** [comment]
I'm receiving the same error. Philips AC2889 model. I can get a ping from the IP address for the device so I know the IP is correct.

Let me know how I can troubleshoot if at all.

**Comment by Sunoo** [comment]
I’m hoping to have something available for you guys to try soon. Life has been a bit more crazy lately than expected, so I haven’t had as much time as I’d like to devote to this.

**Comment by lumaxis** [comment]
@Sunoo Is there anything concrete people can help you with? I just got an AC2889 and would be happy to assist with e.g. capturing some of the network traffic, if that would help.

**Comment by Sunoo** [comment]
At the moment, I just need a little bit more time. I plan to be publishing a plugin within the next few days that should have at least partial support for the new protocol. Sorry for the delays.

**Comment by lumaxis** [comment]
Yea, no worries at all! Just trying to figure out ways to support you, take your time 🙂

**Comment by Sunoo** [comment]
Development would go a bit faster if I had access to one of the new units, but I think I’ve wrapped my head around the new protocol enough to have a shot at it. Plus the py-air-ctrl project seems to have done a fairly good job of documenting it.

**Comment by lumaxis** [comment]
Oh, sweet. That project does look helpful indeed.
Well, if there's anything anyone can help you with, please do let us know 👍

**Comment by lumaxis** [comment]
@Sunoo Thanks to the virus, I have plenty of time on my hands right now. If you want to push your current state somewhere, I'd be happy to test/contribute!

**Comment by Sunoo** [comment]
Sure, let me see about getting it up shortly. Looking like (at least for now) it makes sense to have this as a separate project than this one, because they’re almost completely different. I may look into merging them in the future, for clarity to end users.

**Comment by dvd2000k** [comment]
Hi, I'm interested too. Is there any update? Thank you for your work.

**Comment by kaydee1982** [comment]
If it helps, I've just got a new AC1214, yesterday and it has the same problem.
Thanks for your time and effort you put in this plugin!

**Comment by jednl** [comment]
Hi, I’ve got the same issue on a new AC5659.

**Comment by dsadsa897897r** [comment]
They seem to be making good progress over here:

https://github.com/rgerganov/py-air-control/issues/35

I'm getting this error though - any ideas:

Unable to load status info: SyntaxError: Unexpected token í in JSON at position 0

**Comment by Sunoo** [comment]
I've been watching that project, it seems pretty close to having the new units figured out. As soon as they release their new version, I plan to resume work and get a Homebridge plugin together for them. I apologize for the delay.

**Comment by Cyber1000** [comment]
@dsadsa897897r did you try the master of this repo or my repo at  https://github.com/Cyber1000/py-air-control/tree/coap_1_0_7?
you could use the -d switch to get some more info (or if you are lucky the broken JSON as simple string)

**Comment by dsadsa897897r** [comment]
This was with this repo - I only got the JSON error once in the homebridge logs and it fixed itself - so no idea sorry! But all working well now! Need to find a good way of turning on and off the blue light on the front though - I currently call a script to do it and it's ugly!

**Comment by Cyber1000** [comment]
ah ok, since you mentioned the other project and an error in nearly the same sentence, I mixed that up 😄 
Thanks for clarifying!

**Comment by Sunoo** [comment]
> This was with this repo - I only got the JSON error once in the homebridge logs and it fixed itself - so no idea sorry! But all working well now! 

Glad it seems to have resolved itself. If it happens again, open a new issue here and I’ll do my best to help get to the bottom of it. I had also understood this to be in reference to the other project, which is why I didn’t respond sooner.

> Need to find a good way of turning on and off the blue light on the front though - I currently call a script to do it and it's ugly!

“Night mode” controls that light, which HomeKit didn’t provide an easy way to expose in their air purifier device. I could add a config option that exposes a light switch device that could toggle night mode? Would that work for your setup?

**Comment by dsadsa897897r** [comment]
Thank you - I will let you know if the error appears again.

Ah - I did not know about Night Mode! 

Exposing the device as a light switch would be great for my particular setup.

**Comment by Sunoo** [comment]
@dsadsa897897r Okay! I’ll try to get to it this weekend, but no promises.

**Comment by dsadsa897897r** [comment]
that would be amazing!

**Comment by Sunoo** [comment]
It looks like the py-air-control project has gotten the new units working reliably. I plan to spend some time soon, probably this weekend, getting a version of my plugin together that makes use of their updates. I’ll post here once I have something.

**Comment by Woodstock37** [comment]
Thanks for your update, I get an unit model#AC3033/30, but I have no luck to make it works, and here is the log:

[6/1/2020, 10:49:35 AM] [philipsAir] Initializing platform accessory Living Room Purifier...
[6/1/2020, 10:49:36 AM] [philipsAir] Unable to load key: FetchError: request to http://192.168.1.128/di/v1/products/0/security failed, reason: connect ECONNREFUSED 192.168.1.128:80
[6/1/2020, 10:49:36 AM] [philipsAir] Unable to load firmware info: FetchError: request to http://192.168.1.128/di/v1/products/0/firmware failed, reason: connect ECONNREFUSED 192.168.1.128:80
[6/1/2020, 10:49:36 AM] [philipsAir] Unable to load status info: FetchError: request to http://192.168.1.128/di/v1/products/1/air failed, reason: connect ECONNREFUSED 192.168.1.128:80
[6/1/2020, 10:49:36 AM] [philipsAir] Unable to load filter info: FetchError: request to http://192.168.1.128/di/v1/products/1/fltsts failed, reason: connect ECONNREFUSED 192.168.1.128:80

Any idea to get it fix? Thanks.

**Comment by Sunoo** [comment]
Sorry, but I haven’t added support for the new models yet, which is why you’re getting the error. I’m hoping to get something in place soon.

**Comment by Woodstock37** [comment]
Thanks for your reply and your hard work. Look forward to have this model to be supported soon.

**Comment by zsiraph** [comment]
AC4558/50 (4500i) user, sadly I got the same error as Woodstock37.

I do understand that the support for the new models are coming, and I want to thank you for your amazing and hard work!

Cheers!

**Comment by Sunoo** [comment]
It’s basically just a matter of free time and motivation right now. I was planning to work on it this past weekend, but everything going on in the world made it difficult to focus enough to make meaningful progress. I apologize for the delays.

**Comment by Sunoo** [comment]
Okay, good news everyone! I've just pushed an update that should support all new models. Note that there are some additional steps in the installation instructions, which you'll need to follow even if upgrading.

As I've mentioned before, I do not own or have access to any of these newer models, so I'm relying on you guys to let me know if anything doesn't work as expected.

Sorry this took so long, but I was relying on others to figure out the protocols used.

**Comment by zsiraph** [comment]
Hello!
Great news! 
However, sadly it does not work for me, the device is added to Home, but it is not accessable.

All I got from the log is this:

`2020. 06. 15. 6:42:43 [Philips Air Cleaner] Initializing platform accessory Living Room Air Cleaner...
2020. 06. 15. 6:42:43 [Philips Air Cleaner] Unable to load firmware info: TypeError: Cannot read property 'name' of undefined
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:43 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:43 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:43 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:43 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:43 self.send(msg)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:43 self.connect()
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:43 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:43 raise err
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:43 sock.connect(sa)
2020. 06. 15. 6:42:43 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:43 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 191, in _get
2020. 06. 15. 6:42:43 return self._get_once(url)
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 184, in _get_once
2020. 06. 15. 6:42:43 with urllib.request.urlopen(url) as response:
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:43 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:43 response = self._open(req, data)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:43 '_open', req)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:43 result = func(*args)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:43 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:43 raise URLError(err)
2020. 06. 15. 6:42:43 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:43 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:43 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:43 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:43 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:43 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:43 self.send(msg)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:43 self.connect()
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:43 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:43 raise err
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:43 sock.connect(sa)
2020. 06. 15. 6:42:43 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:43 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 210, in get_firmware
2020. 06. 15. 6:42:43 firmware = self._get(url)
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 195, in _get
2020. 06. 15. 6:42:43 self._get_key()
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 116, in _get_key
2020. 06. 15. 6:42:43 with urllib.request.urlopen(req) as response:
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:43 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:43 response = self._open(req, data)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:43 '_open', req)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:43 result = func(*args)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:43 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:43 raise URLError(err)
2020. 06. 15. 6:42:43 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:43 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:43 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:43 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:43 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:43 self.send(msg)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:43 self.connect()
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:43 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:43 raise err
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:43 sock.connect(sa)
2020. 06. 15. 6:42:43 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:43 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 191, in _get
2020. 06. 15. 6:42:43 return self._get_once(url)
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 184, in _get_once
2020. 06. 15. 6:42:43 with urllib.request.urlopen(url) as response:
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:43 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:43 response = self._open(req, data)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:43 '_open', req)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:43 result = func(*args)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:43 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:43 raise URLError(err)
2020. 06. 15. 6:42:43 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:43 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:43 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:43 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:43 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:43 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:43 self.send(msg)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:43 self.connect()
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:43 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:43 raise err
2020. 06. 15. 6:42:43 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:43 sock.connect(sa)
2020. 06. 15. 6:42:43 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:43 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:43 Traceback (most recent call last):
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 200, in get_status
2020. 06. 15. 6:42:43 status = self._get(url)
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 195, in _get
2020. 06. 15. 6:42:43 self._get_key()
2020. 06. 15. 6:42:43 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 116, in _get_key
2020. 06. 15. 6:42:43 [Philips Air Cleaner] Unable to load status info: TypeError: Cannot read property 'mode' of undefined
2020. 06. 15. 6:42:43 [Philips Air Cleaner] Unable to load filter info: TypeError: Cannot read property 'fltsts0' of undefined
2020. 06. 15. 6:42:43 [GreeAC]: connecting to 192.168.50.11 [using source port 8011]
2020. 06. 15. 6:42:43 Bridge is running on port 51826.
2020. 06. 15. 6:42:43 [GreeAC] New device added: 1ed46a88 - 192.168.50.11
2020. 06. 15. 6:42:43 [GreeAC] device is bound: 1ed46a88 - v6Xy9Ab2De5Gh8Jk
2020. 06. 15. 6:42:43 [ZsirapH-AC] Connected to 1ed46a88 with IP address 192.168.50.11
2020. 06. 15. 6:42:48 [Ring] Configuring 1 cameras and 2 devices for location "Pécs" - locationId: b95cef62-c35b-4294-aa65-45409717e94a
2020. 06. 15. 6:42:48 [Ring] Adding new accessory doorbell_v4 Front Door
2020. 06. 15. 6:42:48 Front Door is running on port 43953.
2020. 06. 15. 6:42:54 with urllib.request.urlopen(req) as response:
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:54 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:54 response = self._open(req, data)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:54 '_open', req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:54 result = func(*args)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:54 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:54 raise URLError(err)
2020. 06. 15. 6:42:54 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:54 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:54 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:54 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:54 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:54 self.send(msg)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:54 self.connect()
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:54 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:54 raise err
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:54 sock.connect(sa)
2020. 06. 15. 6:42:54 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:54 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 191, in _get
2020. 06. 15. 6:42:54 return self._get_once(url)
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 184, in _get_once
2020. 06. 15. 6:42:54 with urllib.request.urlopen(url) as response:
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:54 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:54 response = self._open(req, data)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:54 '_open', req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:54 result = func(*args)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:54 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:54 raise URLError(err)
2020. 06. 15. 6:42:54 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:54 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:54 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:54 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:54 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:54 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:54 self.send(msg)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:54 self.connect()
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:54 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:54 raise err
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:54 sock.connect(sa)
2020. 06. 15. 6:42:54 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:54 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 215, in get_filters
2020. 06. 15. 6:42:54 filters = self._get(url)
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 195, in _get
2020. 06. 15. 6:42:54 self._get_key()
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 116, in _get_key
2020. 06. 15. 6:42:54 with urllib.request.urlopen(req) as response:
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:54 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:54 response = self._open(req, data)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:54 '_open', req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:54 result = func(*args)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:54 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:54 raise URLError(err)
2020. 06. 15. 6:42:54 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:54 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:54 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:54 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:54 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:54 self.send(msg)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:54 self.connect()
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 938, in connect
2020. 06. 15. 6:42:54 (self.host,self.port), self.timeout, self.source_address)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/socket.py", line 727, in create_connection
2020. 06. 15. 6:42:54 raise err
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/socket.py", line 716, in create_connection
2020. 06. 15. 6:42:54 sock.connect(sa)
2020. 06. 15. 6:42:54 ConnectionRefusedError: [Errno 111] Connection refused
2020. 06. 15. 6:42:54 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 191, in _get
2020. 06. 15. 6:42:54 return self._get_once(url)
2020. 06. 15. 6:42:54 File "/usr/local/lib/python3.7/dist-packages/pyairctrl/http_client.py", line 184, in _get_once
2020. 06. 15. 6:42:54 with urllib.request.urlopen(url) as response:
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
2020. 06. 15. 6:42:54 return opener.open(url, data, timeout)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 525, in open
2020. 06. 15. 6:42:54 response = self._open(req, data)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
2020. 06. 15. 6:42:54 '_open', req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
2020. 06. 15. 6:42:54 result = func(*args)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
2020. 06. 15. 6:42:54 return self.do_open(http.client.HTTPConnection, req)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1319, in do_open
2020. 06. 15. 6:42:54 raise URLError(err)
2020. 06. 15. 6:42:54 urllib.error.URLError: <urlopen error [Errno 111] Connection refused>
2020. 06. 15. 6:42:54 During handling of the above exception, another exception occurred:
2020. 06. 15. 6:42:54 Traceback (most recent call last):
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/urllib/request.py", line 1317, in do_open
2020. 06. 15. 6:42:54 encode_chunked=req.has_header('Transfer-encoding'))
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1244, in request
2020. 06. 15. 6:42:54 self._send_request(method, url, body, headers, encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
2020. 06. 15. 6:42:54 self.endheaders(body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
2020. 06. 15. 6:42:54 self._send_output(message_body, encode_chunked=encode_chunked)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
2020. 06. 15. 6:42:54 self.send(msg)
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 966, in send
2020. 06. 15. 6:42:54 self.connect()
2020. 06. 15. 6:42:54 File "/usr/lib/python3.7/http/client.py", line 938, in connect`

**Comment by Sunoo** [comment]
Looks like you are using the HTTP protocol, newer models use CoAP or Plain CoAP. Have you tried those yet?

**Comment by zsiraph** [comment]
> Looks like you are using the HTTP protocol, newer models use CoAP or Plain CoAP. Have you tried those yet?

Thanks, trying right now! All I have to do is use "protocol": "coap" instead of "protocol": "http" in the config, right?

**Comment by Sunoo** [comment]
Correct, the options for that field are http, plain_coap, and coap. At this time I don’t know any way to detect the correct protocol without just trying each one and seeing what works.

**Comment by zsiraph** [comment]
> Correct, the options for that field are http, plain_coap, and coap. At this time I don’t know any way to detect the correct protocol without just trying each one and seeing what works.

Great, that fixed it!
Can you help me with the "sleep_speed" and "light_control" settings? What variables should I use with those parameters? 

I have an AC4558/50 (4500i), which does support sleep mode and has light controls.

Thanks!

**Comment by Sunoo** [comment]
Fantastic! Glad it’s working for you!

If it supports both those features, set those values to true, no quotes needed.

**Comment by zsiraph** [comment]
> Fantastic! Glad it’s working for you!
> 
> If it supports both those features, set those values to true, no quotes needed.

Thanks! However, light control is not working as I thought -at least with my device - (the bulbs are in Homekit, but the device does not turn off lights despite the fact that I turned them off in the Home app), but that is not a big issue for me, I just wanted to let you know as a feedback.

There are 3 air quality "stats" for the AC4558/50 (4500i):

- PM 2.5
- Gas
- Indoor Allergen index

The Air quality tile is showing PM 2.5 Density and Air Quality right now, but no Gas and Indoor Allergen. Is there any way to get those in the Air quality tile? Also it looks like that the PM Density and Air Quality values are not updating correctly, since the numbers are not the same in the home app tile and on the device screen.

Thanks again for this plugin, it is amazing!

**Comment by Sunoo** [comment]
It’s possible lights work differently on these new models. IAI is mapped to Air Quality, it was the best solution I could come up with. I’m not sure about Gas, my unit doesn’t have that. I’ll see if I can find out more about it, and will support it if I can.

**Comment by Sunoo** [comment]
I looked into it, I don’t see gas as a value coming back, so it is most likely being calculated based on some other values.

As for the lights, it looks like all units should be controlled the same. Perhaps your unit just doesn’t allow control of them?

**Comment by lumaxis** [comment]
Thanks @Sunoo! Looking forward to trying this out. How do I know which protocol to use with which device model?

**Comment by Sunoo** [comment]
Unfortunately, I’m not aware of any way to tell, other than just trial and error. Choosing the wrong protocol won’t break anything though.
